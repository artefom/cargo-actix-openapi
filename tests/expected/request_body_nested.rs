//! API auto-generated by apigen

use std::fmt::Display;

use std::{collections::HashMap, fmt::Debug};

use serde::{Deserialize, Serialize};

use actix_web::{
    get, http::StatusCode, middleware::NormalizePath, web, App, HttpRequest, HttpResponse,
    HttpServer, ResponseError,
};

use async_trait::async_trait;

// Defaults
// -------------------------------
fn default_int_1() -> i64 {
    1
}
fn default_float_0_1() -> f64 {
    0.1
}

// Enums
// -------------------------------

// Struct
// -------------------------------

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
pub struct GreetUserPath {
    /// The name of the user to greet.
    pub user: String,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
pub struct GreetUserBodyObj {
    #[serde(default = "default_int_1")]
    pub foo: i64,
    #[serde(default = "default_float_0_1")]
    pub bar: f64,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
pub struct GreetUserBody {
    pub str: String,
    pub obj: GreetUserBodyObj,
}

// Error with details
// -------------------------------


/// Create detailed errors with ease
#[macro_export]
macro_rules! detailed {
    ($err:expr,$msg:expr) => {
        $crate::server::api::Detailed {
            error: $err,
            details: $msg.to_string(),
        }
    };
}

/// Bails with detailed api error
#[macro_export]
macro_rules! apibail {
    ($err:expr,$msg:expr) => {
        return Err($crate::server::api::Detailed {
            error: $err,
            details: $msg.to_string(),
        })
    };
}

pub trait StatusCoded {
    fn status_code(&self) -> StatusCode;
}

#[derive(Debug)]
pub struct Detailed<E> {
    pub error: E,
    pub details: String,
}

impl<E: Display> Display for Detailed<E> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}. Reason: {}", self.error, self.details)
    }
}

impl<E: Display + Debug> std::error::Error for Detailed<E> {}

impl<E: Display + Debug> ResponseError for Detailed<E>
where
    E: StatusCoded,
{
    fn status_code(&self) -> StatusCode {
        self.error.status_code()
    }
}

/// Converts some result to Result with detailed api error
pub trait ApiErr<T, E> {
    /// Wrap the error value with additional context.
    fn apierr<C>(self, err: C) -> Result<T, Detailed<C>>
    where
        C: Display + Send + Sync + 'static;
}

impl<T, E> ApiErr<T, E> for Result<T, E>
where
    E: Debug + Send + Sync + 'static,
{
    fn apierr<C>(self, err: C) -> Result<T, Detailed<C>>
    where
        C: Display + Send + Sync + 'static,
    {
        // Not using map_err to save 2 useless frames off the captured backtrace
        // in ext_context.
        match self {
            Ok(ok) => Ok(ok),
            Err(original_error) => Err(Detailed {
                error: err,
                details: format!("{:?}", original_error),
            }),
        }
    }
}

// Error
// -------------------------------

// Api service
// -------------------------------

#[async_trait(?Send)]
pub trait ApiService<S>
where
    S: Send + Sync + 'static,
{
    /// Returns a greeting to the user!
    async fn greet_user(
        data: web::Data<S>,
        path: web::Path<GreetUserPath>,
        body: Option<web::Json<GreetUserBody>>,
    ) -> web::Json<String>;
}

// Run service function (+ helper functions)
// -----------------------------------------
static DOCS_OPENAPI: &str = include_str!("static/openapi.yaml");
static DOCS_HTML: &str = include_str!("static/docs.html");
async fn openapi() -> String {
    DOCS_OPENAPI.to_string()
}
async fn docs() -> HttpResponse {
    HttpResponse::build(StatusCode::OK)
        .content_type("text/html; charset=utf-8")
        .body(DOCS_HTML)
}
async fn to_docs() -> HttpResponse {
    HttpResponse::build(StatusCode::TEMPORARY_REDIRECT)
        .append_header(("Location", "v1/docs"))
        .body("")
}
async fn to_docs_noprefix() -> HttpResponse {
    HttpResponse::build(StatusCode::TEMPORARY_REDIRECT)
        .append_header(("Location", "docs"))
        .body("")
}

pub async fn run_service<T, S>(bind: &str, initial_state: S) -> Result<(), std::io::Error>
where
    T: ApiService<S> + 'static,
    S: Send + Sync + 'static,
{
    let app_data = web::Data::new(initial_state);

    use web::{get,post,delete};

    HttpServer::new(move || {
        App::new()
            .app_data(app_data.clone())
            .wrap(NormalizePath::trim())
            // Static paths
            .route("/", get().to(to_docs_noprefix))
            .route("/docs", get().to(docs))
            .route("/openapi.yaml", get().to(openapi))
            .route("/v1", get().to(to_docs))
            .route("/v1/docs", get().to(docs))
            .route("/v1/openapi.yaml", get().to(openapi))
            // Server routes
            .route("/hello/{user}", post().to(T::greet_user))
            .route("/v1/hello/{user}", post().to(T::greet_user))
    })
    .bind(bind)?
    .run()
    .await?;

    Ok(())
}