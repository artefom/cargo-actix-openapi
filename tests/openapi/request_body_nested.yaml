openapi: "3.0.3"
info:
  version: "1.0"
  title: "Hello World API"
paths:
  /hello/{user}:
    parameters:
      - name: user
        in: path
        required: true
        description: The name of the user to greet.
        schema:
          type: string

    post:
      operationId: greet_user
      summary: Returns a greeting to the user!
      description: Some detailed description
      requestBody:
        content:
          application/json:
            schema:
              type: object
              required:
                - str
                - obj
              properties:
                str:
                  type: string
                obj:
                  type: object
                  properties:
                    foo:
                      type: integer
                      default: 1
                    bar:
                      type: number
                      default: 0.1
      responses:
        200:
          description: Returns the greeting.
          content:
            application/json:
              schema:
                type: string
