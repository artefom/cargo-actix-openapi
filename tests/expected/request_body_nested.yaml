api:
  definitions:
    DOCS_OPENAPI:
      data: !StaticStr
        path: static/openapi.yaml
    DOCS_HTML:
      data: !StaticStr
        path: static/docs.html
    openapi:
      data: !StaticStringPath
        data: DOCS_OPENAPI
    docs:
      data: !StaticHtmlPath
        data: DOCS_HTML
    GreetUserPath:
      data: !Struct
        doc: null
        properties:
        - name: user
          rename: user
          default: null
          type_: String
          doc: The name of the user to greet.
    default_int_1:
      data: !DefaultProvider
        vtype: i64
        value: '1'
    default_float_0_1:
      data: !DefaultProvider
        vtype: f64
        value: '0.1'
    GreetUserBodyObj:
      data: !Struct
        doc: null
        properties:
        - name: foo
          rename: foo
          default: default_int_1
          type_: i64
          doc: null
        - name: bar
          rename: bar
          default: default_float_0_1
          type_: f64
          doc: null
    GreetUserBody:
      data: !Struct
        doc: null
        properties:
        - name: str
          rename: str
          default: null
          type_: String
          doc: null
        - name: obj
          rename: obj
          default: null
          type_: GreetUserBodyObj
          doc: null
  operations:
    greet_user:
      doc: Returns a greeting to the user!
      param_path: web::Path<GreetUserPath>
      param_query: null
      param_body: Option<web::Json<GreetUserBody>>
      response: web::Json<String>
  paths:
  - operation: greet_user
    path: /hello/{user}
    method: Post
  - operation: greet_user
    path: /v1/hello/{user}
    method: Post
  static_services:
  - method: Get
    path: /openapi.yaml
    data: openapi
  - method: Get
    path: /docs
    data: docs
  - method: Get
    path: /v1/openapi.yaml
    data: openapi
  - method: Get
    path: /v1/docs
    data: docs
